### PR TITLE
DOC: stats: probability tutorial/transition guide

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Toward gh-15928

#### What does this implement/fix?
This is a ~~draft of a~~ guide to support transition between the old and new infrastructures for working with probability distributions.

#### Additional information
~~I plan to add quite a bit more about fitting. For now, I'd like to see how this renders so I can determine whether/how to break it up.~~ Done.

### CODE REVIEW

**Comment by mdhaber** [comment]
@melissawm can you show me how to add images (e g. PNG) in the markdown notebook tutorial?

**Comment by mdhaber** [comment]
I'll try to finish this up tonight.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [inline doc/source/tutorial/stats/rv_infrastructure.md:528]: What's the rationale for having `X = stats.Normal()` but not `X = stats.Gamma()`?

**Review by h-vetinari — COMMENTED** [review]
Please take this only take this as a comment from the peanut gallery - the work on stats has been so sprawling and extensive that it would take a long time to dig this up.

It's also not a criticism in any way, shape, or form; just trying to understand the context.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:445]: From the top:

> I plan to add quite a bit more about fitting.

Maybe if I haden't done made all the improvements to the existing `fit` this would have been an easier sell. Alas.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:341]: It wasn't removed. Users are welcome to use `expect` from the old interface. It hasn't been a priority for me to add to the new interface because be really can't do much to improve upon the existing numerical integration features (or those we have recently added). Faced with the choice between spending time recreating these interfaces vs adding value, I chose the latter. Also, sometimes users seem to expect more from `expect` than it can reasonably provide (e.g. gh-8928, gh-9259, gh-14347). If `expect` were not an option, perhaps users will learn to perform numerical integration with the general tools. At their simplest, the general tools are not any more complex to use, but they give you a lot more control when things don't "just work".

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:528]: For the _first_ distribution with a new class, `Normal` seemed appropriate. It's pretty fundamental.

Personally, I've envisioned adding a few of the most fundamental distributions back with the new naming scheme because the performance, accuracy, and interface can be a bit better than if we rely on `make_distribution` for them. For instance, `Uniform` can be parametrized by the endpoints rather than location/scale, and new options for private methods can improve the speed and accuracy of the public ones. As for why that's not *all* done yet, well, it's mostly a matter of time. `make_distribution` barely made it into 1.15, and that seemed more important to me, so that's where I spent my time.

**Comment by tylerjereddy** [comment]
Since this is docs only I'll just suggest keeping the backport label and we can put it in before SciPy `1.15.0` RC2 goes out. I'd like to aim for getting RC1 out today since I've blocked off the time and it is a bit overdue.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md]: I made some improvements. I would like to expand with examples of fitting with other objective functions and entirely different requirements (e.g. users have asked to fit a distribution not to *data* but such that specific statistics assume particular values), but I think it is much more convincing. I didn't even consciously choose this case to be problematic for MLE, but it happens to be a great example of the sort of pitfals I've seen in "bug" reports.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [inline doc/source/tutorial/stats/rv_infrastructure.md:441]:
```suggestion
Another problem is that MLE is inherently problematic for some distributions. In this example, for instance, we can make the NLLF as low as we please simply by setting the location to the minimum of the data - even for bogus values of the shape and scale parameters.
```

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [inline doc/source/tutorial/stats/rv_infrastructure.md:508]: Hah I got distracted. But this is where I was going to show that we can easily adjust the objective function.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [reply doc/source/tutorial/stats/rv_infrastructure.md:575]: > As for why that's not all done yet, well, it's mostly a matter of time. make_distribution barely made it into 1.15, and that seemed more important to me, so that's where I spent my time.

Completely understandable. Might still be worth a thought to note that the new infrastructure will still receive updates & improvements in the . The text kinda reads like it's in its final state already.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [inline doc/source/tutorial/stats/rv_infrastructure.md:914]:
```suggestion
- Additional built-in distributions (besides `Normal` and `Uniform`)
- Discrete distributions
```

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:648]: Hah, that would be nice. Please see if the (new) conclusion begins to address this point, and let me know if there are particular places where the language might be changed to avoid the appearance that we think this is all done.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [inline doc/source/tutorial/stats/rv_infrastructure.md:64]: 👍

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [inline doc/source/tutorial/stats/rv_infrastructure.md:72]:
```suggestion
In the new infrastructure, distributions families are classes named according to `CamelCase` conventions (note that not all distributions have a corresponding class yet; see `make_distribution` below). They must be instantiated before use, with parameters passed as keyword-only arguments.
```

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [inline doc/source/tutorial/stats/rv_infrastructure.md:20]:
```suggestion
```{eval-rst}
.. jupyterlite:: ../../_contents/rv_infrastructure.ipynb
   :new_tab: True
```

(rv_infrastructure)=
# Random Variable Transition Guide
```

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [inline doc/source/tutorial/stats/rv_infrastructure.md:156]: The default plot title is a bit awkward `Normal(mu=np.float64(0.0), sigma=np.float64(1.0))`. We could add a title manually; in an case, this is not a requirement, just a comment.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [inline doc/source/tutorial/stats/rv_infrastructure.md:475]: Using params both as a 2-tuple and 3-tuple in the same codeblock is confusing IMO. I would keep the custom `nllf` as using a 3-typle, and then just discard `loc` when unpacking `params`:
```
    c, _, scale = params
```

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [inline doc/source/tutorial/stats/rv_infrastructure.md:515]: Could you add a comment what acronym `nlps` stands for?

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [inline doc/source/tutorial/stats/rv_infrastructure.md:561]: `gh-12134` does not automatically create an issue-link; we should make it clickable

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [inline doc/source/tutorial/stats/rv_infrastructure.md:596]: The legends for the plots in this section are off (presumably should show both a red line and a dashed green line, rather than one blue one).

image

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [inline doc/source/tutorial/stats/rv_infrastructure.md:670]: This should be a code cell.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [inline doc/source/tutorial/stats/rv_infrastructure.md:904]: Very nice!

**Review by h-vetinari — COMMENTED** [review]
I saw you pushed a new change but didn't want to redo my already in-progress review 😅 

Before I started reviewing, the doc build failed with:
```
/home/circleci/repo/build-install/lib/python3.11/site-packages/scipy/stats/_new_distributions.py:docstring of scipy.stats._new_distributions.Normal:81: WARNING: Failed to create a cross reference. A title or caption not found: 'rv_infrastructure'
/home/circleci/repo/build-install/lib/python3.11/site-packages/scipy/stats/_new_distributions.py:docstring of scipy.stats._new_distributions.Uniform:81: WARNING: Failed to create a cross reference. A title or caption not found: 'rv_infrastructure'

Exited with code exit status 1
```
Other than that and some minor suggestions below, this looks very good to me already! Thanks a lot for this huge body of work!

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:156]: Actually if @tylerjereddy will allow a backport for this, this is a little regression in the rendering. In a few lines, @steppi could add `__str__` overrides very similar to the existing `__repr__` overrides that would fix this.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:475]: I will try to make that more clear. It will need to be just 2-tuple in the optimization, though.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:515]: Sure. In the meantime, "negative log product spacing".

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:561]: Sure, I forgot that one.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:596]: Yeah, I'll play with it. Thanks.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:670]: Weird... I prepared all this as a notebook. Maybe by accident I made the cell markdown instead of code after writing it.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:72]: 👍 I put something like this a little here.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [reply doc/source/tutorial/stats/rv_infrastructure.md:162]: Since it looks doc-only, I'm in favour of doing that.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (h-vetinari) [reply doc/source/tutorial/stats/rv_infrastructure.md]: I would still prefer that the `params` in the signature of the (here only local) `nllf` matches the 3-tuple params in the rest of the notebook, rather than introducing a 2-tuple variant just for this code-block

Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md]: To make sure we're talking about the same thing, it's not just in this code block: there are only two parameters being optimized, so all the code blocks that define an object function do this. 

So we would have to show `loc` being added as a decision variable and then constrained to be equal to 0. Many solvers have trouble with that. We happened to have given some `minimize` solvers the ability to handle that, but not all, so I would rather avoid show that. 

Can you think of other things that would address your concern? For example, usually in objective functions, the variable is just called `x`. I can change the name of the argument of objective functions to just `x`?

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (steppi) [reply doc/source/tutorial/stats/rv_infrastructure.md:162]: @mdhaber, would rather I push a commit with the `__str__` overrides here, or make a separate PR against `main`?

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (steppi) [reply doc/source/tutorial/stats/rv_infrastructure.md]: If we want to be completely unambiguous we could use `decision_variables` instead of `params`. That should make sure no one would mistakenly assume that `params` had some specific meaning other than just the variables being optimized.

Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md]: Hmm, but that is still a bit ambiguous because there is a distinction between the variables themselves and the value at that iteration. Maybe `values_of_decision_variables_at_this_iteration`?

Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md]: Actually, I had already removed three-element use of `params` before the last commit. I was just immediately unpacking calls to `weibull_min.fit` into `c_, loc_, scale_`. I went ahead and removed the remaining uses of `params`, though, for good measure.

**Comment by mdhaber** [comment]
gh-22081 makes the plots look much better!

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:162]: Resolved in gh-22081

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [inline doc/source/tutorial/stats/rv_infrastructure.md:92]:
```suggestion
The documentation of [`scipy.stats.Normal`](https://scipy.github.io/devdocs/reference/generated/scipy.stats.Normal.html#scipy.stats.Normal) contains links to detailed documentation of each of its methods. (Compare, for instance, against the documentation of [`scipy.stats.norm`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.norm.html).)
```

**Comment by mdhaber** [comment]
With gh-22081 merged, I'm marking this ready for review again. I'll resolve the (easy) CI failure and (trivial) merge conflict when there are other comments to address. The rendered document is [here](https://output.circle-artifacts.com/output/job/d4382e47-0c39-4fed-835f-17468039da30/artifacts/0/html/tutorial/stats/rv_infrastructure.html).

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (j-bowhay) [reply doc/source/tutorial/stats/rv_infrastructure.md:895]: Also possibly 
```suggestion
The `plot` method is quite flexible, with a signature inspired by grammar of graphics. For instance, with the argument `x` on `[-10, 10]`, plot the `pdf` against the `cdf`.
```

Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md:895]:
```suggestion
The `plot` method is quite flexible, with a signature inspired by the [Grammar of Graphics](https://books.google.com/books?id=_kRX4LoFfGQC&printsec=frontcover&source=gbs_ge_summary_r&cad=0#v=onepage&q&f=false). For instance, with the argument `x` on $[-10, 10]$, plot the PDF against the CDF.
```

**Review by steppi — APPROVED** [review]
This looks great @mdhaber. Flows well. Just one minor wording suggestion. This has my approval, but I'll leave it open for comments from others even after Matt fixes the merge conflicts and applies suggestions.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md]: 807b3cb2057b8452b8a8329ca369466cacc3f474

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [reply doc/source/tutorial/stats/rv_infrastructure.md]: 807b3cb2057b8452b8a8329ca369466cacc3f474

**Review by j-bowhay — APPROVED** [review]
Looks great!

**Comment by mdhaber** [comment]
Hmm looks like I need to do whatever was done in gh-22042 before merge.

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [inline doc/source/tutorial/stats/rv_infrastructure.md:14]:
```suggestion
+++ {"tags": ["jupyterlite_sphinx_strip"]}

```{eval-rst}
```

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [inline doc/source/tutorial/stats/rv_infrastructure.md:20]:
```suggestion
(rv_infrastructure)=

+++

# Random Variable Transition Guide
```

**Review by mdhaber — COMMENTED** [review]
See gh-22042

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [inline doc/source/tutorial/stats/rv_infrastructure.md:437]:
```suggestion
First, note that MLE is one of several potential strategies for fitting distributions to data. It does nothing more than find the values of the distribution parameters for which the negative log-likelihood (NLLF) is minimized. Note that for the distribution from which the data were generated, the NLLF is:
```

File: doc/source/tutorial/stats/rv_infrastructure.md
Comment (mdhaber) [inline doc/source/tutorial/stats/rv_infrastructure.md:449]:
```suggestion
Therefore, `fit` considers its job done, whether or not this satisfies the user's notions of a "good fit" or whether the parameters are within reasonable bounds. Besides allowing the user to specify a guess for distribution parameters, it gives little control over the process. (*Note: technically, the `optimizer` argument can be used for greater control, but this is not well documented, and its use negates any convenience offered by the `fit` method.*)
```

**Review by mdhaber — COMMENTED** [review]
A few things I'd like to adjust with `[skip ci]`.

@steppi, it's been a few days. What do you think about committing this and merging so I can update the release notes with the link?

File: doc/source/tutorial/stats/probability_distributions.rst
Comment (mdhaber) [inline doc/source/tutorial/stats/probability_distributions.rst:7]:
```suggestion
advantages. For the new infrastructure, see* :doc:`rv_infrastructure`.
```

**Comment by steppi** [comment]
> @steppi, it's been a few days. What do you think about committing this and merging so I can update the release notes with the link?

Sure. Pushing the button

**Comment by tylerjereddy** [comment]
Please try to squash-merge backports

**Pull request merged**
